packages/ens-referrals/src/referrer.ts

```
@@ -0,0 +1,98 @@
+
+/**
+ * Encoded Referrer byte length
+ *
+ * The count of bytes the {@link EncodedReferrer} value consists of.
+ */
+export const ENCODED_REFERRER_BYTE_LENGTH = 32;
+
+/**
+ * Encoded Referrer Padding
+ *
+ * The initial bytes of correctly encoded referrer value.
+ */
+export const encodedReferrerPadding = pad("0x", {
+  size: ENCODED_REFERRER_BYTE_OFFSET,
+  dir: "left",
+});
+
+/**
+ * Zero Encoded Referrer
+ *
+ * Guaranteed to be a hex string representation of a 32-byte zero value.
+ */
+export const zeroEncodedReferrer = pad("0x", {
+  size: ENCODED_REFERRER_BYTE_LENGTH,
+  dir: "left",
+});
+
+/**
+ * Build an {@link EncodedReferrer} value for the given {@link Address}
+ * according to the subjective referrer encoding used for ENS Holiday Awards.
+ */
+export function buildEncodedReferrer(address: Address): EncodedReferrer {
+  const lowercaseAddress = address.toLowerCase() as Address;
+
+  return pad(lowercaseAddress, { size: ENCODED_REFERRER_BYTE_LENGTH, dir: "left" });
+}
+
+/**
+ * Decode an {@link EncodedReferrer} value into a checksummed {@link Address}
+ * according to the subjective referrer encoding used for ENS Holiday Awards.
+ *
+ * @param encodedReferrer - The "raw" {@link EncodedReferrer} value to decode.
+ * @returns The decoded referrer checksummed address.
+ */
+export function decodeEncodedReferrer(encodedReferrer: Hex): Address {
+  // return zero address if encoded referrer is not of expected size
+  if (size(encodedReferrer) !== ENCODED_REFERRER_BYTE_LENGTH) {
+    return zeroAddress;
+  }
+
+  const padding = slice(encodedReferrer, 0, ENCODED_REFERRER_BYTE_OFFSET);
+
+  // return zero address if the padding of encoded referrer is not correct
+  if (padding !== encodedReferrerPadding) {
+    return zeroAddress;
+  }
+
+  const decodedReferrer = slice(encodedReferrer, ENCODED_REFERRER_BYTE_OFFSET);
+
+  try {
+    // return checksummed address
+    return getAddress(decodedReferrer);
+  } catch {
```

Hmm, I believe that both of the following cases this function should throw an error (and document this is JSDocs) rather than return zeroAddress:

1. `encodedReferrer` is incorrect length.
2. `getAddress` throws an error.

If either of these fail, doesn't that mean that the EncodedReferrer passed to this function does not meet the guarantees of an EncodedReferrer? And assuming so, it should throw an error rather than silently return zeroAddress.

We must respect the type system and our type definitions.